Demo On-Chain Messaging

mehtaishita opened the pull request:
### Goal

The goal of this PR is to create a demo for registering avro schemas and submitting messages.

**Closes** https://github.com/LibertyDSNP/example-client-frequency/issues/5

### Discussion
Currently the demo uses a hard coded avro schema. In the future story, would like to add input field for inputting a JSON representation of an avro schema and do all the same things that it does right now. 

### Checklist

- [x] Imported avsc for avro schema serialization

mehtaishita committed:
main setup

errors out - but pieces in place

mehtaishita committed:
main setup

errors out - but pieces in place

mehtaishita committed:
update version

mehtaishita committed:
wip

mehtaishita committed:
wip

mehtaishita committed:
removal

mehtaishita committed:
removal

mehtaishita committed:
removal

mehtaishita committed:
removal

mehtaishita committed:
basic button

mehtaishita committed:
basic button

mehtaishita committed:
rewrite

register static schema

mehtaishita committed:
rewrite

register static schema

mehtaishita committed:
hardcoded functions - fetch, submit

mehtaishita committed:
hardcoded functions - fetch, submit

mehtaishita committed:
fix extrinsics

mehtaishita committed:
fix extrinsics

mehtaishita committed:
add and get messages

mehtaishita committed:
add and get messages

mehtaishita committed:
npm pkg

mehtaishita committed:
npm pkg

mehtaishita committed:
fix

mehtaishita committed:
fix

mehtaishita committed:
callback for list

mehtaishita committed:
callback for list

mehtaishita committed:
no set

mehtaishita committed:
no set

mehtaishita committed:
final

mehtaishita committed:
final

mehtaishita committed:
avro

mehtaishita committed:
avro

mehtaishita committed:
ver

mehtaishita committed:
bye

mehtaishita committed:
avsc

mehtaishita committed:
avsc

mehtaishita committed:
webpack errors

2 hours of webpack issues with avsc later...

mehtaishita committed:
webpack errors

2 hours of webpack issues with avsc later...

mehtaishita committed:
ignore lock file

mehtaishita committed:
ignore lock file

mehtaishita committed:
fixing package types

wip

mehtaishita committed:
fixing package types

wip

mehtaishita committed:
wip - react ui

mehtaishita committed:
wip - react ui

mehtaishita committed:
rpc failing to be found

mehtaishita committed:
rpc failing to be found

mehtaishita committed:
use state in react

mehtaishita committed:
use state in react

mehtaishita committed:
why am I getting a compile error on api types

mehtaishita committed:
why am I getting a compile error on api types

wilwade committed:
Fix issue with ExtrinsicStatus

wilwade committed:
Fix issue with ExtrinsicStatus

wilwade committed:
Update api-augment package to fix types

wilwade committed:
Update api-augment package to fix types

wilwade committed:
Remove all the any

wilwade committed:
Remove all the any

mehtaishita committed:
yes

mehtaishita committed:
yes

mehtaishita committed:
config plus main

mehtaishita committed:
config plus main

mehtaishita committed:
Merge branch 'Demo/on-chain-messaging/287' of github.com:LibertyDSNP/example-client-frequency into Demo/on-chain-messaging/287

# Conflicts:
#	src/services/chain/apis/common.ts

mehtaishita committed:
func name change

mehtaishita committed:
func name change

mehtaishita committed:
remove logs

mehtaishita committed:
remove logs

mehtaishita committed:
list schemas

mehtaishita committed:
list schemas

mehtaishita committed:
some lint

mehtaishita committed:
some lint

mehtaishita committed:
very basic...pulled back everything else

mehtaishita committed:
very basic...pulled back everything else

mehtaishita committed:
w

mehtaishita committed:
w

mehtaishita committed:
css

mehtaishita committed:
css

mehtaishita committed:
what I had for demo

mehtaishita committed:
what I had for demo

mehtaishita committed:
removing commented out code

mehtaishita committed:
removing commented out code

mehtaishita committed:
pull out schema component

mehtaishita committed:
pull out schema component

mehtaishita committed:
moving static schemas to componenet

mehtaishita committed:
moving static schemas to componenet

mehtaishita committed:
pulled out schemas componenet + styles

mehtaishita committed:
pulled out schemas componenet + styles

mehtaishita committed:
fetch messages - parameter

mehtaishita committed:
fetch messages - parameter

mehtaishita committed:
create message component

mehtaishita committed:
create message component

mehtaishita committed:
base components for messages etc

mehtaishita committed:
base components for messages etc

mehtaishita committed:
re submitting yarn

mehtaishita committed:
re submitting yarn

mehtaishita committed:
update react

mehtaishita committed:
update react

mehtaishita committed:
trying to create a dynamic from from schema json

mehtaishita committed:
trying to create a dynamic from from schema json

mehtaishita committed:
Revert "trying to create a dynamic from from schema json"

This reverts commit 21299706e1ee320a5b07ebf03aec4c55150704cd.

mehtaishita committed:
Revert "trying to create a dynamic from from schema json"

This reverts commit 21299706e1ee320a5b07ebf03aec4c55150704cd.

mehtaishita committed:
create message things

mehtaishita committed:
create message things

mehtaishita committed:
quick feedback from register schema

mehtaishita committed:
quick feedback from register schema

mehtaishita committed:
validate message set up

mehtaishita committed:
validate message set up

mehtaishita committed:
create message validates and submits message

mehtaishita committed:
create message validates and submits message

mehtaishita committed:
lodash

mehtaishita committed:
lodash

mehtaishita committed:
lodash start case

mehtaishita committed:
lodash start case

mehtaishita committed:
moving schema section after wallet

mehtaishita committed:
moving schema section after wallet

mehtaishita committed:
react table component for schemas

mehtaishita committed:
react table component for schemas

mehtaishita committed:
removing console log

mehtaishita committed:
removing console log

mehtaishita committed:
cleaning up main

mehtaishita committed:
cleaning up main

mehtaishita committed:
cleaning up console log

mehtaishita committed:
cleaning up console log

mehtaishita committed:
console log

mehtaishita committed:
console log

mehtaishita committed:
refactor - schema details type

mehtaishita committed:
refactor - schema details type

mehtaishita committed:
list of schemas table - render columns properly

mehtaishita committed:
list of schemas table - render columns properly

mehtaishita committed:
missing comma

mehtaishita committed:
missing comma

mehtaishita committed:
selected row stored in state

something's wrong with it - the proper key isn't being stored

mehtaishita committed:
selected row stored in state

something's wrong with it - the proper key isn't being stored

mehtaishita committed:
proper schema id returned

mehtaishita committed:
proper schema id returned

mehtaishita committed:
renaming type file

mehtaishita committed:
renaming type file

mehtaishita committed:
refactor - message details type, list of messages table setup

mehtaishita committed:
refactor - message details type, list of messages table setup

mehtaishita committed:
updating import

mehtaishita committed:
updating import

mehtaishita committed:
passing state from child to parent to child

It's not the prettiest - the props definition for schema and messages are messy

mehtaishita committed:
passing state from child to parent to child

It's not the prettiest - the props definition for schema and messages are messy

mehtaishita committed:
small table + expand rows

and also no props definition for schema componenet

mehtaishita committed:
small table + expand rows

and also no props definition for schema componenet

mehtaishita committed:
updating types lodash

mehtaishita committed:
updating types lodash

mehtaishita committed:
wip - decode/encode message

buffer gets truncated

mehtaishita committed:
wip - decode/encode message

buffer gets truncated

mehtaishita committed:
moving create message and list message as child components to list of schemas

mehtaishita committed:
moving create message and list message as child components to list of schemas

mehtaishita committed:
renaming message props to schema props

mehtaishita committed:
renaming message props to schema props

mehtaishita committed:
wip - message serialization

buffer keeps getting truncated

mehtaishita committed:
wip - message serialization

buffer keeps getting truncated

mehtaishita committed:
updated frequency

mehtaishita committed:
updated frequency

mehtaishita committed:
list all messages - map message response to message details

mehtaishita committed:
list all messages - map message response to message details

mehtaishita committed:
toggle both messages components on click button

mehtaishita committed:
toggle both messages components on click button

mehtaishita committed:
update types and clean up extrinsics

mehtaishita committed:
update types and clean up extrinsics

mehtaishita committed:
remove unused styles

mehtaishita committed:
remove unused styles

mehtaishita committed:
sticking to the one test schema

mehtaishita committed:
sticking to the one test schema

mehtaishita committed:
create message component respects 'is visible' bool

mehtaishita committed:
create message component respects 'is visible' bool

mehtaishita committed:
using antd components

mehtaishita committed:
using antd components

mehtaishita committed:
rename button

mehtaishita committed:
rename button

mehtaishita committed:
schemas table - actions are not buttons

mehtaishita committed:
schemas table - actions are not buttons

mehtaishita committed:
remove unused import

mehtaishita committed:
remove unused import

mehtaishita committed:
running yarn format

I'm only pushing the format I ran on my code - I don't want to confuse this with linting changes for code I didn't touch this time - for review purposes

mehtaishita committed:
running yarn format

I'm only pushing the format I ran on my code - I don't want to confuse this with linting changes for code I didn't touch this time - for review purposes

mehtaishita committed:
register multiple schemas

mehtaishita committed:
register multiple schemas

mehtaishita committed:
using only one hardcoded schema

mehtaishita committed:
using only one hardcoded schema

mehtaishita commented on src/components/CreateMessage.tsx:
```diff
@@ -0,0 +1,78 @@
+        let avroBuffer: Buffer = staticSchema.toBuffer(form.getFieldsValue());
+        console.log("avro buffer: ", avroBuffer);
+        await addMessage(avroBuffer, parseInt(props?.schema?.schema_id));
+    }
```
Validating and submitting messages to hard coded 'staticSchema'

mehtaishita commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,49 @@
+            setListOfMessage(allMessages);
+    }
+
+    if (props.isVisible)
```
component is visible based on the props parameter, as set by the parent component

mehtaishita commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,49 @@
+
+    if (props.isVisible)
+        {return ( <>
+            <Button onClick={listMessages}>Refresh List</Button>
```
Using 'Refresh List' button here to trigger fetching the messages. In the future, would be better to 'useEffect' so the data refreshes as soon as the component loads. I didn't do that right now because it takes a while, and I couldn't get a 'Loading Spinner' to work

mehtaishita commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,46 @@
+            <Column title= 'Model Type'dataIndex='model_type' key= 'model_type'
+                    render={(type) => upperFirst(lowerCase(type)) }/>
+            <Column title= 'Payload Location' dataIndex='payload_location' key= 'payload_location'
+                    render={(loc) => upperFirst(lowerCase(loc))} />
```
Title case from camel case

wilwade replied:
Might be worth writing a small function for this instead of using lodash:

```
const upperFirstLowerRest = (str: string) => {
  const head = (str[0] || "").toUpperCase();
  const rest = str.slice(1).toLowerCase();
  return head + rest;
}

mehtaishita commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,46 @@
+                render={(record) => ( <>
+                      <Space size="middle">
+                        <a onClick={() => {setSelectedSchema(record); setShowCreateMessageComp(!showCreateMessageComp)}}>Create Message</a>
+                        <a onClick={() => {setSelectedSchema(record); setShowListMessageComp(!showLlistMessageComp)}}>List Messages</a>
```
here passing in selected schema, so the message can be submitted for the current row's schema id

mehtaishita commented on src/services/chain/apis/extrinsic.ts:
```diff
@@ -116,4 +119,67 @@ export const createMsaForProvider = async (callback: DsnpCallback,
+    payload_location: payload_location.toString(),
+    model_structure: modelParsed
+    }
+};
```
This formatted result is probably not ideal in the extrinsic - but as we add more schemas to the demo, will refactor

mehtaishita commented:
**Choices made for the sake of demo/time:**

- Using hard coded avro schema
- Therefore using a hard coded form to validate and submit messages for the above schema

**Future Story / refactor:** 

- Input field for Json representation of an avro schema
- Dynamically build the form for submitting the message as per Avro schema's fields
- Input field for Json representation of a message
- Refactor shared components into smaller chunks (such as table for Listing schemas and Listing messages)

mehtaishita committed:
spacing between components

mehtaishita committed:
spacing between components

wilwade commented on src/App.tsx:
```diff
@@ -1,8 +1,6 @@
-import React, {useEffect, useState} from 'react';
```
I think we still need `import React from 'react';` right?

mehtaishita replied:
technically, no. It's an unused import

wilwade replied:
Ah I see. That's a new feature of React 17 https://www.chakshunyu.com/blog/why-did-and-dont-you-need-to-import-react/

wilwade commented on src/components/CreateMessage.tsx:
```diff
@@ -0,0 +1,78 @@
+
+    const handleSubmit =  async () => {
+
+        let isValid = validateMessage();
+        if (isValid) {
```
I'd likely collapse this.

```suggestion
        if (validateMessage()) {
```

wilwade commented on src/components/CreateMessage.tsx:
```diff
@@ -0,0 +1,78 @@
+        if (isValid) {
+            submitMessage();
+        };
+        form.resetFields();
```
Do we want to reset if the validation fails?

mehtaishita replied:
changed to behavior to simply write text 'message not valid' without resetting the fields

wilwade reviewed (commented):
Got through part of it. Also there is a rebase that you will need to do

mehtaishita committed:
newest version

mehtaishita committed:
newest version

mehtaishita committed:
matching version to current one

mehtaishita committed:
matching version to current one

mehtaishita committed:
Merge branch 'Demo/on-chain-messaging/287' of github.com:LibertyDSNP/example-client-frequency into Demo/on-chain-messaging/287

# Conflicts:
#	src/services/chain/polkadot.ts

mehtaishita committed:
renmaed frequency chain api augment

mehtaishita committed:
validation message message

mehtaishita committed:
removed it by mistake

wilwade commented on package.json:
```diff
@@ -3,33 +3,47 @@
   "version": "0.1.0",
```
We should not have a package-lock.json file as this uses yarn

rlaferla replied:
That's my fault but please remove it.

wilwade commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,48 @@
+                    </Space> </>
+            )} />
+        </Table>
+        <CreateMessage schema={selectedSchema} isVisible={showCreateMessageComp}/>
+        <ListMessages schema={selectedSchema} isVisible={showLlistMessageComp}/>
```
In React, visibility is usually handled this way:

```suggestion
        {showCreateMessageComp && <CreateMessage schema={selectedSchema}/>}
        {showLlistMessageComp && <ListMessages schema={selectedSchema}/>}
```

That way it only shows if `show...Comp` is true and you don't have to have the show/hide logic in the component.

wilwade commented on src/services/types.ts:
```diff
@@ -0,0 +1,21 @@
+import { Json, u32 } from "@polkadot/types";
+import React from "react";
+
+export interface SchemaDetails {
```
I find these are usually put in the same file as the component, but I might be out of date on my react patterns.

mehtaishita replied:
I'm also using SchemaDetails in extrinsic.ts so figured it was best to leave it separate?

wilwade replied:
Ah that's fine then.

wilwade commented on src/services/types.ts:
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+
+export interface SchemaDetails {
+  key: React.Key;
```
I don't think you need this `key` field

mehtaishita replied:
If I don't use the key, I get: 
**react.development.js:209 Warning: Each child in a list should have a unique "key" prop.**

wilwade replied:
Ah I see. I guess that works. You could also just use [Table](https://ant.design/components/table/#API) `rowKey` to then map it to `schema_id`. Either way is fine.

wilwade commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,51 @@
+                <Column
+                    title="Messages"
+                    dataIndex="payload"
+                    key="payload"
```
I don't think you need this `key` field because the dataIndex is unique

wilwade commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,51 @@
+                <Column
+                    title="Message Length"
+                    dataIndex="payload_length"
+                    key="payload_length" />
```
I don't think you need this `key` field because the dataIndex is unique

wilwade commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,51 @@
+
+        let allMessages: MessageDetails[] = messages.map((msg) => {
+            return {
+                key: msg.block_number.toString(),
```
If you need this key, it doesn't work. Key needs to be unique and block_number is not (well it is only on instant sealing

You likely want to use the index of the map function to give it a unique key number.

wilwade reviewed (changes requested):
Part 2. Almost through it

mehtaishita committed:
visibility boolean

mehtaishita committed:
extra import

mehtaishita committed:
removing lodash + short fn for camel case to title case

mehtaishita committed:
updating frequency api augment

mehtaishita committed:
unique key for messages

mehtaishita committed:
updated yarn

wilwade commented on package.json:
```diff
@@ -48,5 +61,9 @@
     ]
+  },
+  "devDependencies": {
+    "@types/lodash": "^4.14.185",
```
We can drop lodash types as well.

wilwade commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,48 @@
+
+        <Table dataSource={listOfSchemas} size="small" expandedRowRender={record => <pre>{JSON.stringify(record.model_structure, null, 2)}</pre>}>
+            <Column title= 'Schema Id' dataIndex= 'schema_id' key= 'schema_id' />
+            <Column title= 'Model Type'dataIndex='model_type' key= 'model_type'
```
Column only needs a key if the `dataIndex` is not unique.
```suggestion
            <Column title= 'Schema Id' dataIndex= 'schema_id' />
            <Column title= 'Model Type'dataIndex='model_type'
```

wilwade commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,48 @@
+            <Column title= 'Schema Id' dataIndex= 'schema_id' key= 'schema_id' />
+            <Column title= 'Model Type'dataIndex='model_type' key= 'model_type'
+                    render={(type) => camelCaseToTitleCase(type) }/>
+            <Column title= 'Payload Location' dataIndex='payload_location' key= 'payload_location'
```
Column only needs a key if the `dataIndex` is not unique.
```suggestion
            <Column title= 'Payload Location' dataIndex='payload_location'
```

wilwade commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,48 @@
+                    render={(loc) => camelCaseToTitleCase(loc)} />
+            <Column
+                title="Messages"
+                key="messages"
```
Column only needs a key if the `dataIndex` is not unique.
```suggestion
```

wilwade commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,50 @@
+    const listMessages = async () => {
+        const messages: MessageResponse[] = await fetchMessagesForSchema(parseInt(props.schema.schema_id));
+
+        let allMessages: MessageDetails[] = messages.map((msg) => {
+            return {
+                key: Number(msg.index) + 1,
```
The index on msg is the index inside the block number. So combining them like: msg.index + "_" + msg.block_number would be unique, or you can just use the index from the map:

```suggestion
        let allMessages: MessageDetails[] = messages.map((msg, index) => {
            return {
                key: index,
```

wilwade commented on src/services/types.ts:
```diff
@@ -0,0 +1,20 @@
+  key: React.Key;
+  payload: Json;
+  payload_length: string;
+}
+
+export interface SchemaProps {
```
This one I would say remove all together however. See comments on the two components that use it.

wilwade commented on src/components/CreateMessage.tsx:
```diff
@@ -0,0 +1,82 @@
+
+const {Text} = Typography;
+
+const CreateMessage = (props: SchemaProps): JSX.Element => {
```
I suggest always having distinct props per component.
```suggestion
const CreateMessage = (props: { schema: SchemaDetail }): JSX.Element => {
```

Or

```suggestion
interface CreateMessageProps {
  schema: SchemaDetails;
}
const CreateMessage = (props: CreateMessageProps): JSX.Element => {
```

wilwade commented on src/components/ListMessages.tsx:
```diff
@@ -0,0 +1,50 @@
+import { MessageDetails, SchemaProps } from "../services/types";
+import { staticSchema } from "./RegisterSchema";
+
+const ListMessages = (props: SchemaProps): JSX.Element => {
```
Here as well. I don't think we would want to share props.

TS allows types to auto cast when they have the same signature. So a variable of type `interface X { x: bool }` can be used on a function with with a argument of type `interface NotX {x: bool }`

wilwade commented on src/components/ListSchemas.tsx:
```diff
@@ -0,0 +1,48 @@
+        setListOfSchemas(schemas);
+    }
+
+    return <div>
```
Actually you don't need this at all, but `<>` over a `div`.
```suggestion
    return <>
```

wilwade reviewed (changes requested):
Part 3

mehtaishita committed:
pr comments

- removed lodash ref
- removed schema props, and replaced with component specific props
- replaced <div> with <>

mehtaishita committed:
key not needed on column type

mehtaishita committed:
message payload length not applicable to on chain messages

mehtaishita committed:
format

mehtaishita commented:
why does github always accept the incoming yarn.lock? How do I resolve the 236 conflicts on yarn.lock without doing it by hand?

mehtaishita committed:
Merge remote-tracking branch 'origin/main' into Demo/on-chain-messaging/287

# Conflicts:
#	package.json
#	yarn.lock

mehtaishita committed:
re added yarn lock

mehtaishita committed:
removing package lock since we use yarn lock

wilwade reviewed (commented):
👍
